doc/developer/design/20230210_storage_persist_sink_redux.md

```diff
@@ -0,0 +1,195 @@
+
+- It is needlessly expensive to create a `BatchBuilder` for each timestamp.
+  - A change to this implementation is to avoid spilling data to persist for timestamps
+  whose values use a small amount of memory. These can be consolidated into large batches in `append_batches`,
+  similar to how <https://github.com/MaterializeInc/materialize/pull/17526> works.
+  - We do this in the current implementation, so doing it from many workers, instead of 1,
+  which has strictly more throughput (though does cause additional `SET` and `GET` load on AWS S3).
+  - Its undecided whether or not this should be required for the reference implementation.
+- `append_batches` and `write_batches` will be required to track statistics to expose to users.
+  - See <https://www.notion.so/materialize/User-facing-metrics-for-Sources-9603b61224a04a749969702e04662ed3> for more details
+
+
+# Rollout
+[rollout]: #rollout
+
+## Testing and observability
+[testing-and-observability]: #testing-and-observability
+
+This new implementation can be tested entirely using the existing CI, nightly, and release-qualification pipelines.
+However its important that the performance of the new implementation is carefully monitored as part of the first
+release that contains it.
+
+## Lifecycle
+[lifecycle]: #lifecycle
+
+This implementation can be validated before release, and does not need any feature gating.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The primary drawback of this design, as alluded to above, is that it does not represent the final implementation we want the
+storage `persist_sink` to have. Ideally, we would be able to use the exact same implementation between compute and storage, but
+a after a _large_ amount of brainstorming within the storage team, this is no feasible without significant engineering effort
+(the technical details here are mentioned in [Conclusion and alternatives](#conclusion-and-alternatives), which we are unable
+to invest at this time.
+
+## Conclusion and alternatives
+[conclusion-and-alternatives]: #conclusion-and-alternatives
+
+Some rejected alternatives are:
+
+- Eliminate the storage `persist_sink` and replace with it with the compute one. Rejected for the following reasons:
+  - Requires skipping self-correction on startup, which is deemed too complex to maintain.
+  - No clear way to keep the self-correction mechanism while also using bounded amounts of memory.
+    - A core design principle of storage is that the ingested collections ("sources") can be
+    arbitrarily sized, and its memory usage should only scale with the size of the largest
+    transaction. The `perist_sink` implementation must conform to this requirement.
+    - Its possible we re-evaluate if the compute sink can use bounded memory in the future,
+    particularly when we have out-of-band consolidation of batches.
+  - Self-correction does additional work that uses compute resources, whereas storage wants to optimize for the
+  raw performance of storage ingestion.
```

Comment:
Couldn't we just connect the corrections source to an empty input that never produces any data? This way, the machinery should be unchanged. I see that it'd still maintain a vector of corrections, but it seems to be equivalent of passing the data itself around.

Reply:
The problem here is that this doesn't prevent us from needing to load all the data for single timestamps into memory. The storage team thought about how to get this mode to work cleanly, using bounded memory, without breaking the compute mode, and it remains unsolved